util/argo/diff.go

```diff
@@ -0,0 +1,379 @@
+package argo
+
+import (
+	"fmt"
+
+	"github.com/go-logr/logr"
+
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	appv1 "github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/argoproj/argo-cd/v2/util/argo/managedfields"
+	appstatecache "github.com/argoproj/argo-cd/v2/util/cache/appstate"
+
+	"github.com/argoproj/gitops-engine/pkg/diff"
+	"github.com/argoproj/gitops-engine/pkg/utils/kube"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// DiffConfigBuilder is used as a safe way to create valid DiffConfigs.
+type DiffConfigBuilder struct {
+	diffConfig *diffConfig
+}
+
+// NewDiffConfigBuilder create a new DiffConfigBuilder instance.
+func NewDiffConfigBuilder() *DiffConfigBuilder {
+	return &DiffConfigBuilder{
+		diffConfig: &diffConfig{},
+	}
+}
+
+// WithIgnores sets the list of ResourceIgnoreDifferences in the diff config. Will set an
+// empty list if the input parameter is nil.
+func (b *DiffConfigBuilder) WithIgnores(i []v1alpha1.ResourceIgnoreDifferences) *DiffConfigBuilder {
+	ignores := i
+	if ignores == nil {
+		ignores = []v1alpha1.ResourceIgnoreDifferences{}
+	}
+	b.diffConfig.ignores = ignores
+	return b
+}
+
+// WithOverrides sets the map of ResourceOverride in the diff config. Will set an
+// empty map if the input parameter is nil.
+func (b *DiffConfigBuilder) WithOverrides(o map[string]v1alpha1.ResourceOverride) *DiffConfigBuilder {
+	overrides := o
+	if overrides == nil {
+		overrides = make(map[string]appv1.ResourceOverride)
+	}
+	b.diffConfig.overrides = overrides
+	return b
+}
+
+// WithAppLabelKey sets the appLabelKey in the diff config.
+func (b *DiffConfigBuilder) WithAppLabelKey(l string) *DiffConfigBuilder {
+	b.diffConfig.appLabelKey = &l
+	return b
+}
+
+// WithTrackingMethod sets the trackingMethod in the diff config.
+func (b *DiffConfigBuilder) WithTrackingMethod(t string) *DiffConfigBuilder {
+	b.diffConfig.trackingMethod = &t
+	return b
+}
+
+// WithAppName sets the appName in the diff config. The appName is only required
+// if retrieving the diff from the cache (NoCache = false).
+func (b *DiffConfigBuilder) WithAppName(n string) *DiffConfigBuilder {
+	b.diffConfig.appName = &n
+	return b
+}
+
+// WithNoCache sets the nocache in the diff config. Defines if it should retrieve
+// the computed diff from the cache.
+func (b *DiffConfigBuilder) WithNoCache(c bool) *DiffConfigBuilder {
+	b.diffConfig.noCache = &c
+	return b
+}
+
+// WithStateCache sets the appstatecache.Cache in the diff config. Only required if
+// retrieving the diff from the cache (NoCache = false).
+func (b *DiffConfigBuilder) WithStateCache(s *appstatecache.Cache) *DiffConfigBuilder {
```

Comment:
Instead of providing `*appstatecache.Cache` and `appName` you can provide `cache retriever` of type (`func(res *[]*appv1.ResourceDiff) error`):

```golang
WithStateCache(func(res *[]*appv1.ResourceDiff) error {
	return m.cache.GetAppManagedResources(app.Name res)
})
```

Not a big deal but might make it easier to test and minus one field in DiffConfig

Reply:
I introduced 2 different methods in the builder instead:
- `WithNoCache()`
- `WithCache(s *appstatecache.Cache, appName string)`

I think this is cleaner and more intuitive for consumers to use. One thing I disliked about the cache retriever idea is that it would mutate the provided `*[]*appv1.ResourceDiff` parameter which would be very confusing.